modules/apps/27-interchain-accounts/host/keeper/relay.go

```diff
@@ -74,7 +74,6 @@ func (k Keeper) executeTx(ctx sdk.Context, sourcePort, destPort, destChannel str
 	}
 
 	// NOTE: The context returned by CacheContext() creates a new EventManager, so events must be correctly propagated back to the current context
-	ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
```

Comment:
Note the safety here is that the cached context is being used to cache execution of all msg executions. We use another cache context for each individual message execution (which emits events using result events since message handler will construct a new event manager) 
ie
```go
cacheCtx, writeCacheCtx:= ctx.CacheContext()
for msg := rangs msgs {
    cacheCtxInner, writeCacheCtxInner:= cacheCtx.CacheContext()
    
    res, err := handler(msg) // does not set emitted events in cacheCtxInner
    if err != nil {
       // all changes discarded
    }
    
    emitEvents(res.Events()) // emit events otherwise not written
    writeCacheContextInner() 
}

writeCacheCtx() // writes events
```

Reply:
++